File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/TelemetryService.kt
package com.windsurf.controller.services

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.SystemInfo
import com.windsurf.controller.GeneratedBuildConfig
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * PostHog telemetry for the JetBrains plugin. Mirrors the VS Code +
 * CLI services (apps/vsc-plugin/src/services/telemetry.service.ts,
 * apps/cli/src/services/telemetry.service.ts) so events land in the
 * same PostHog project under a unified person record.
 *
 * Captured context:
 *   - super properties: pluginVersion, ideName, ideVersion, platform, arch
 *   - person properties (on identify): userId, email, name, plan
 *   - per-event properties: per-call only — never prompt text, never
 *     tokens, never file paths inside the user's project.
 *
 * Opt-out:
 *   - `CODEAM_TELEMETRY=0` env var (parity with CLI / VS Code).
 *   - User's "Help → Data Sharing" privacy policy preference (see
 *     IDE's `PrivacyPolicy` API). Honoured implicitly: if the key is
 *     not baked into the build, capture is a no-op.
 *
 * Transport: plain HTTPS POST to PostHog's /capture/ endpoint — no
 * SDK dependency needed beyond OkHttp (already in the plugin's
 * dependency graph).
 */
@Service(Service.Level.APP)
class TelemetryService {

    private val logger = Logger.getInstance(TelemetryService::class.java)
    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    @Volatile
    private var distinctId: String? = null
    @Volatile
    private var identified: Boolean = false
    @Volatile
    private var initialised: Boolean = false

    fun init(machineId: String) {
        if (initialised) return
        if (isOptedOut()) return
        if (GeneratedBuildConfig.POSTHOG_API_KEY.isEmpty()) return
        distinctId = "jetbrains-$machineId"
        initialised = true
    }

    private fun isOptedOut(): Boolean {
        if (System.getenv("CODEAM_TELEMETRY") == "0") return true
        if (System.getenv("NO_TELEMETRY") != null) return true
        return false
    }

    private fun superProperties(): JsonObject = JsonObject().apply {
        addProperty("pluginVersion", GeneratedBuildConfig.PLUGIN_VERSION)
        addProperty("pluginSurface", "jetbrains")
        addProperty("ideName", System.getProperty("idea.platform.prefix") ?: "Idea")
        addProperty("ideVersion", System.getProperty("idea.version") ?: "unknown")
        addProperty("platform", when {
            SystemInfo.isMac -> "darwin"
            SystemInfo.isWindows -> "win32"
            SystemInfo.isLinux -> "linux"
            else -> "other"
        })
        addProperty("arch", System.getProperty("os.arch") ?: "unknown")
    }

    data class IdentifyParams(
        val userId: String,
        val email: String? = null,
        val name: String? = null,
        val plan: String? = null,
    )

    fun identify(params: IdentifyParams) {
        if (!initialised) return
        val current = distinctId ?: return
        if (identified) return
        identified = true
        // Alias the anonymous machine-scoped id to the real userId so
        // pre-pair events stitch to the now-identified person.
        if (current != params.userId) {
            postBatch(listOf(aliasPayload(current, params.userId)))
        }
        distinctId = params.userId
        val props = JsonObject()
        params.email?.let { props.addProperty("email", it) }
        params.name?.let { props.addProperty("name", it) }
        params.plan?.let { props.addProperty("plan", it) }
        postBatch(listOf(identifyPayload(params.userId, props)))
    }

    /**
     * Fire-and-forget event capture. Runs the HTTP POST on a pooled
     * thread so we don't block the EDT (or the caller). Property
     * names matching token/secret/api-key patterns are dropped before
     * send to defend against future call sites passing a secret.
     */
    fun capture(event: String, properties: Map<String, Any?> = emptyMap()) {
        if (!initialised) return
        val id = distinctId ?: return
        val sanitized = JsonObject()
        for ((k, v) in properties) {
            if (v == null) continue
            if (k.matches(Regex("(?i)token|secret|api[-_]?key|password"))) continue
            when (v) {
                is String -> sanitized.addProperty(k, v)
                is Number -> sanitized.addProperty(k, v)
                is Boolean -> sanitized.addProperty(k, v)
                else -> sanitized.addProperty(k, v.toString())
            }
        }
        ApplicationManager.getApplication().executeOnPooledThread {
            postBatch(listOf(capturePayload(id, event, sanitized)))
        }
    }

    /** Drop a caught error onto PostHog with a trimmed stack. */
    fun captureError(event: String, error: Throwable, extra: Map<String, Any?> = emptyMap()) {
        val stack = error.stackTrace.take(12).joinToString("\n") { it.toString() }
        capture(event, extra + mapOf("errorMessage" to (error.message ?: error::class.qualifiedName), "errorStack" to stack))
    }

    private fun capturePayload(distinct: String, event: String, props: JsonObject): JsonObject {
        val merged = superProperties()
        for (entry in props.entrySet()) merged.add(entry.key, entry.value)
        return JsonObject().apply {
            addProperty("event", event)
            addProperty("distinct_id", distinct)
            add("properties", merged)
            addProperty("timestamp", Instant.now().toString())
        }
    }

    private fun aliasPayload(from: String, to: String): JsonObject = JsonObject().apply {
        addProperty("event", "\$create_alias")
        addProperty("distinct_id", to)
        add("properties", JsonObject().apply {
            addProperty("distinct_id", to)
            addProperty("alias", from)
        })
        addProperty("timestamp", Instant.now().toString())
    }

    private fun identifyPayload(distinct: String, props: JsonObject): JsonObject = JsonObject().apply {
        addProperty("event", "\$identify")
        addProperty("distinct_id", distinct)
        add("\$set", props)
        add("properties", superProperties())
        addProperty("timestamp", Instant.now().toString())
    }

    private fun postBatch(events: List<JsonObject>) {
        if (events.isEmpty()) return
        val body = JsonObject().apply {
            addProperty("api_key", GeneratedBuildConfig.POSTHOG_API_KEY)
            add("batch", com.google.gson.JsonArray().also { arr -> events.forEach { arr.add(it) } })
        }
        val request = Request.Builder()
            .url("${GeneratedBuildConfig.POSTHOG_HOST}/batch/")
            .post(gson.toJson(body).toRequestBody("application/json".toMediaType()))
            .build()
        try {
            client.newCall(request).execute().close()
        } catch (e: Exception) {
            logger.debug("PostHog capture failed: ${e.message}")
        }
    }

    companion object {
        fun getInstance(): TelemetryService =
            ApplicationManager.getApplication().getService(TelemetryService::class.java)
    }
}

File: apps/vsc-plugin/src/services/telemetry.service.ts
/**
 * PostHog telemetry for the VS Code extension. Mirrors the CLI's
 * `apps/cli/src/services/telemetry.service.ts` so events land in the
 * same PostHog project (US ingestion) under a unified person record.
 *
 * Captured context (matches the CLI):
 *   - super properties: pluginVersion, vscodeVersion, platform, arch
 *   - person properties (on identify): userId, email, name, plan
 *   - per-event properties: sessionId, pluginId, agentId, plus
 *     event-specific fields.
 *
 * NEVER captured (privacy invariant — same list as the CLI):
 *   - pluginAuthToken
 *   - model-provider API keys
 *   - prompt text / agent stdout / file contents / file paths
 *
 * Opt-out:
 *   - VS Code's `telemetry.telemetryLevel` setting set to `off` →
 *     no capture, no init. We respect the user's global IDE setting
 *     in addition to the explicit CodeAgent toggle.
 *   - `codeagent-mobile.telemetryEnabled` configuration setting set
 *     to false.
 *   - `CODEAM_TELEMETRY=0` env var (parity with the CLI).
 */

import * as vscode from 'vscode';
import { PostHog } from 'posthog-node';

// Injected by esbuild's `define` at build time. Empty in dev builds
// unless POSTHOG_API_KEY is set in env when running `npm run build`.
declare const __POSTHOG_API_KEY__: string;
declare const __POSTHOG_HOST__: string;
declare const __PLUGIN_VERSION__: string;

let client: PostHog | null = null;
let distinctId: string | null = null;
let identified = false;

function isOptedOut(): boolean {
  // VS Code's global telemetry switch wins — if the user disabled
  // telemetry IDE-wide, we don't capture anything.
  const ideLevel = vscode.env.isTelemetryEnabled;
  if (ideLevel === false) return true;
  if (process.env.CODEAM_TELEMETRY === '0') return true;
  if (process.env.NO_TELEMETRY) return true;
  const cfg = vscode.workspace.getConfiguration('codeagent-mobile');
  if (cfg.get<boolean>('telemetryEnabled', true) === false) return true;
  return false;
}

type TelemetryProperties = Record<string, string | number | boolean | null | undefined>;
type PostHogJsonProperties = Record<string, string | number | boolean | null>;

function superProperties(): PostHogJsonProperties {
  return {
    pluginVersion: typeof __PLUGIN_VERSION__ !== 'undefined' ? __PLUGIN_VERSION__ : '0.0.0-dev',
    pluginSurface: 'vscode',
    vscodeVersion: vscode.version,
    platform: process.platform,
    arch: process.arch,
  };
}

/**
 * Initialise the PostHog client. Safe to call more than once. Uses
 * `vscode.env.machineId` as the anonymous distinct id — survives
 * across activations + matches the IDE's own telemetry id space.
 * Returns true when capture is active.
 */
export function initTelemetry(): boolean {
  if (client) return true;
  if (isOptedOut()) return false;
  const apiKey = typeof __POSTHOG_API_KEY__ !== 'undefined' ? __POSTHOG_API_KEY__ : '';
  if (!apiKey) return false;
  const host =
    typeof __POSTHOG_HOST__ !== 'undefined' ? __POSTHOG_HOST__ : 'https://us.i.posthog.com';
  distinctId = `vscode-${vscode.env.machineId}`;
  client = new PostHog(apiKey, {
    host,
    flushAt: 20,
    flushInterval: 10_000,
  });
  client.register(superProperties());
  return true;
}

export interface IdentifyParams {
  userId: string;
  email?: string;
  name?: string;
  plan?: string;
}

export function identifyUser(params: IdentifyParams): void {
  if (!client || !distinctId) return;
  if (identified) return;
  identified = true;
  if (distinctId !== params.userId) {
    client.alias({ distinctId, alias: params.userId });
  }
  distinctId = params.userId;
  const personProps: PostHogJsonProperties = {};
  if (params.email !== undefined) personProps.email = params.email;
  if (params.name !== undefined) personProps.name = params.name;
  if (params.plan !== undefined) personProps.plan = params.plan;
  client.identify({ distinctId: params.userId, properties: personProps });
}

/**
 * Fire-and-forget event capture. PostHog batches internally and
 * flushes via the interval — no need to await.
 */
export function capture(event: string, properties: TelemetryProperties = {}): void {
  if (!client || !distinctId) return;
  const safe: PostHogJsonProperties = {};
  for (const [k, v] of Object.entries(properties)) {
    if (v === undefined) continue;
    // Match the CLI's redaction filter: any field name that looks
    // like a secret gets dropped before send. Cheap defense against
    // a future caller passing a token property.
    if (/token|secret|api[-_]?key|password/i.test(k)) continue;
    safe[k] = v;
  }
  client.capture({ distinctId, event, properties: safe });
}

/**
 * Capture a generic error with stack. Use for caught exceptions in
 * fire-and-forget paths where logging is the only signal today —
 * 401-recovery, SSE drop, observer-bridge failures. Stack frames
 * outside `apps/vsc-plugin/` are trimmed to keep PostHog row sizes
 * bounded.
 */
export function captureError(event: string, error: unknown, extra: TelemetryProperties = {}): void {
  const err = error instanceof Error ? error : new Error(String(error));
  const stack = (err.stack ?? '').split('\n').slice(0, 12).join('\n');
  capture(event, { ...extra, errorMessage: err.message, errorStack: stack });
}

export async function shutdownTelemetry(): Promise<void> {
  if (!client) return;
  try {
    await client.shutdown();
  } catch { /* best-effort */ } finally {
    client = null;
    identified = false;
  }
}
